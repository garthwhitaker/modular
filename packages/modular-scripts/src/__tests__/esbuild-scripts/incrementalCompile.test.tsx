import * as esbuild from 'esbuild';
import * as path from 'path';
import * as tmp from 'tmp';
import svgrPlugin from '../../esbuild-scripts/plugins/svgr';
import incrementalCompilePlugin from '../../esbuild-scripts/plugins/incrementalCompile';
import getModularRoot from '../../utils/getModularRoot';
import { Paths } from '../../utils/createPaths';
import prepareUrls from '../../esbuild-scripts/config/urls';

class MockLogger {
  private buffer: string[] = [];
  private started = false;
  start() {
    this.started = true;
  }
  getOutput() {
    return Array.from(this.buffer);
  }
  log(...x: string[]) {
    if (this.started) {
      this.buffer.push(...x);
    }
  }
  clear() {
    this.buffer.push('</CLEAR>');
  }
}

const logger = new MockLogger();
jest.mock('../../utils/logger', () => {
  return {
    __esModule: true,
    get debug() {
      return logger.log.bind(logger);
    },
    get log() {
      return logger.log.bind(logger);
    },
    get clear() {
      return logger.clear.bind(logger);
    },
  };
});

describe('When running a build with incremental compile', () => {
  let tmpDir: tmp.DirResult;
  let outdir: string;
  let compilePromise: Promise<void>;
  let compilePlugin: esbuild.Plugin;
  let complete = false;

  beforeAll(() => {
    tmpDir = tmp.dirSync();
    outdir = path.join(tmpDir.name, 'output');

    const paths = {
      appPackageJson: path.join(getModularRoot(), 'package.json'),
    } as unknown as Paths;
    const urls = prepareUrls('https', 'localhost', 8080);
    const incrementalCompile = incrementalCompilePlugin(paths, urls);
    compilePlugin = incrementalCompile.plugin;
    compilePromise = incrementalCompile.initialBuildPromise;
    void compilePromise.then(() => {
      complete = true;
    });

    logger.start();
  });

  it('SHOULD have a promise which is unresolved', () => {
    expect(complete).toEqual(false);
  });

  describe('WHEN the build is run', () => {
    beforeAll(async () => {
      await esbuild.build({
        entryPoints: [
          path.join(__dirname, '__fixtures__', 'svgr-component', 'index.tsx'),
        ],
        plugins: [svgrPlugin(), compilePlugin],
        // so that we don't get huge output sizes...
        external: ['react', 'react-dom'],
        outdir,
        bundle: true,
        splitting: true,
        format: 'esm',
        target: 'es2021',
      });
    });

    it('SHOULD make the incremental compile promise as done', () => {
      expect(complete).toEqual(true);
    });

    it('SHOULD have logger buffer output', () => {
      expect(logger.getOutput()).toMatchSnapshot();
    });
  });
});
