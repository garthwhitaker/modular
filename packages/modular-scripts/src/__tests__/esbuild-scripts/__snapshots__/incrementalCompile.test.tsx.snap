// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When running a build with incremental compile WHEN the build is run SHOULD have logger buffer output 1`] = `
Array [
  "</CLEAR>",
  "Compiling...",
  "</CLEAR>",
  "[32mCompiled successfully![39m [2m[249ms][22m",
  "You can now view [1mmodular[22m in the browser.",
  "  https://localhost:[1m8080[22m/",
  "Note that the development build is not optimized.",
  "To create a production build, use [36myarn build[39m.",
]
`;
